src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java

```diff
@@ -0,0 +1,390 @@
+                .matchFirstInInventoryWithInventory(
+                        entity,
+                        stack -> stack != null && is.isItemEqual(stack),
+                        InventoryFunctions::doNothing
+                );
+    }
+
+    @Override
+    public void readFromNBT(final NBTTagCompound compound)
+    {
+        super.readFromNBT(compound);
+    }
+
+    @NotNull
+    @Override
+    public NBTTagCompound writeToNBT(@NotNull final NBTTagCompound compound)
+    {
+        super.writeToNBT(compound);
+        return compound;
+    }
+
+    /**
+     * Dump the inventory of a citizen into the warehouse.
+     * Go through all items and search the right chest to dump it in.
+     * @param inventoryCitizen
+     */
+    public void dumpInventoryIntoWareHouse(@NotNull final InventoryCitizen inventoryCitizen)
+    {
+        for(int i = 0; i < inventoryCitizen.getSizeInventory(); i++)
+        {
+            ItemStack stack = inventoryCitizen.getStackInSlot(i);
+
+            if(stack == null || stack.getItem() == null || stack.stackSize == 0)
+            {
+                continue;
+            }
+            TileEntityChest chest = searchRightChestForStack(stack);
+            if(chest == null)
+            {
+                LanguageHandler.sendPlayersLocalizedMessage(getColony().getMessageEntityPlayers(), "com.minecolonies.coremod.wareHouse.full");
+                return;
+            }
+            InventoryUtils.addItemStackToInventory(chest, stack);
+            inventoryCitizen.removeStackFromSlot(i);
+        }
+
+    }
+
+    /**
+     * Search the right chest for an itemStack.
+     * @param stack the stack to dump.
+     * @return the tile entity of the chest
+     */
+    @Nullable
+    private TileEntityChest searchRightChestForStack(@NotNull final ItemStack stack)
+    {
+        if(InventoryUtils.findFirstSlotInInventoryWith(this, stack.getItem(), stack.getItemDamage()) != -1 && InventoryUtils.getOpenSlot(this) != -1)
+        {
+            return this;
+        }
+
+        for(BlockPos pos : getBuilding().getAdditionalCountainers())
+        {
+            TileEntity entity = worldObj.getTileEntity(pos);
+            if(entity instanceof TileEntityChest
+                    && InventoryUtils.findFirstSlotInInventoryWith((TileEntityChest) entity, stack.getItem(), stack.getItemDamage()) != -1
+                    && InventoryUtils.getOpenSlot(this) != -1)
+            {
+                return this;
+            }
+        }
+
+        return searchMostEmptySlot();
+    }
+
+    /**
+     * Search for the chest with the least items in it.
+     * @return the tileEntity of this chest.
+     */
+    @Nullable
+    private TileEntityChest searchMostEmptySlot()
+    {
+        int freeSlots = 0;
+        TileEntityChest emptiestChest = null;
+        for(BlockPos pos : getBuilding().getAdditionalCountainers())
+        {
+            TileEntity entity = worldObj.getTileEntity(pos);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'entity' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)